src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -310,248 +503,354 @@ public void afterSave() {
             throws IOException, InterruptedException {
-        if (event != null) {
-            observer = event.filter(this, observer);
-        }
-        StandardUsernamePasswordCredentials scanCredentials = getScanCredentials();
-        if (scanCredentials == null) {
-            listener.getLogger().format("Connecting to %s with no credentials, anonymous access%n", bitbucketUrl());
-        } else {
-            listener.getLogger().format("Connecting to %s using %s%n", bitbucketUrl(), CredentialsNameProvider.name(scanCredentials));
-        }
-        // this has the side-effect of ensuring that repository type is always populated.
-        listener.getLogger().format("Repository type: %s%n", WordUtils.capitalizeFully(getRepositoryType().name()));
+        try (BitbucketSCMSourceRequest request = new BitbucketSCMSourceContext(criteria, observer)
+                .withTraits(traits)
+                .newRequest(this, listener)) {
+            StandardUsernamePasswordCredentials scanCredentials = credentials();
+            if (scanCredentials == null) {
+                listener.getLogger().format("Connecting to %s with no credentials, anonymous access%n", getServerUrl());
+            } else {
+                listener.getLogger().format("Connecting to %s using %s%n", getServerUrl(),
+                        CredentialsNameProvider.name(scanCredentials));
+            }
+            // this has the side-effect of ensuring that repository type is always populated.
+            listener.getLogger().format("Repository type: %s%n", WordUtils.capitalizeFully(getRepositoryType().name()));
+            // populate the request with its data sources
+            if (request.isFetchPRs()) {
+                request.setPullRequests(new LazyIterable<BitbucketPullRequest>() {
+                    @Override
+                    protected Iterable<BitbucketPullRequest> create() {
+                        try {
+                            return (Iterable<BitbucketPullRequest>) buildBitbucketClient().getPullRequests();
+                        } catch (IOException | InterruptedException e) {
+                            throw new BitbucketSCMSource.WrappedException(e);
+                        }
+                    }
+                });
+            }
+            if (request.isFetchBranches()) {
+                request.setBranches(new LazyIterable<BitbucketBranch>() {
+                    @Override
+                    protected Iterable<BitbucketBranch> create() {
+                        try {
+                            return (Iterable<BitbucketBranch>) buildBitbucketClient().getBranches();
+                        } catch (IOException | InterruptedException e) {
+                            throw new BitbucketSCMSource.WrappedException(e);
+                        }
+                    }
+                });
+            }
+            if (request.isFetchTags()) {
```

Comment:
A marker for future enhancements. I wanted to leave placeholders for where the tag support will go... requires some enhancements to scm-api and branch-api (specifically the ability to declare some SCMHead instances as interesting vs non-interesting... which is wanted for PR building too - think origin PRs, the origin branch becomes non-interesting when it is part of a PR... a non-interesting branch should not be automatically built... that lets you retain the build history of the origin branch while running the PR... which handles the use-case of a long-lived branch with PRs back to master)

Tags would mostly be non-interesting (unless you have a trait to mark some tags as interesting)... thus when you discover a repository, you would get jobs for each tag but they would not build automatically and you can go into Jenkins and manually request a build of any tag that you want to build